SPEC.md

```diff
@@ -0,0 +1,334 @@
+# Nexus RPC HTTP Specification
+
+## Overview
+
+The Nexus protocol, as specified below, is a synchronous RPC protocol. Arbitrary length operations are modelled on top
+of a set of pre-defined synchronous RPCs.
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. The caller can cancel an asynchronous operation, check for its outcome, or fetch its current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation when it
+is ready.
+
+## Operation Addressability
+
+An operation is addressed using three components:
+
+- [Service Name](#service-name)
+- [Operation Name](#operation-name)
+- [Operation ID](#operation-id)
+
+> **Note:** More details regarding IDs will be provided upon finalizing the ID proposal.
+
+### Service Name
+
+Service names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation Name
+
+Operation names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation ID
+
+Operation IDs:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+## Schema Definitions
+
+### Failure
+
+The `Failure` object represents outcomes that are `failed` or `canceled`. The object MUST adhere to the following JSON
+schema:
+
+```yaml
```

Comment:
I was wondering about that, the only time we pass this around it should be accompanied by a `Nexus-Operation-State` header.

Reply:
But state is not the same as failure type/kind. You can have a running state but a failure kind of "not-done-yet" if the wait timed out the way I am thinking about it.

Reply:
Are you also talking about application failure types or just framework level failure types?

Reply:
The failure can be either. In fact, aside from state, we should use this failure object for every single failure we have (including not found, 5xx, etc).

Reply:
I'm leaning towards using 204 with a `Nexus-Operation-State: running` header as shown below (and I may require that successful completion is returned with a 200 code).

I will, however, use this failure object for all non successful HTTP responses.

Reply:
Having a hard time justifying adding a type field here.
This will always come in either application failure context with the proper header or request failure accompanied by an HTTP status code.